sql/src/main/java/io/crate/executor/transport/ExecutionNodesTask.java

```diff
@@ -0,0 +1,304 @@
+        final ContextCloser contextCloser = new ContextCloser(transportCloseContextNodeAction);
+        Futures.addCallback(result, new FutureCallback<TaskResult>() {
+            @Override
+            public void onSuccess(TaskResult result) {
+                closeContext();
+            }
+
+            @Override
+            public void onFailure(@Nonnull Throwable t) {
+                closeContext();
+            }
+
+            private void closeContext() {
+                if (nodesByServer.isEmpty()) {
+                    return;
+                }
+                for (ExecutionNode executionNode : executionNodes) {
+                    contextCloser.process(executionNode, nodesByServer.keySet());
+                }
+            }
+        });
+    }
+
+    private void addStatsTablesCallback(final StatsTables statsTables, final RamAccountingContext ramAccountingContext) {
+        Futures.addCallback(result, new FutureCallback<TaskResult>() {
+            @Override
+            public void onSuccess(TaskResult result) {
+                statsTables.operationFinished(mergeOperationId, null, ramAccountingContext.totalBytes());
+            }
+
+            @Override
+            public void onFailure(@Nonnull Throwable t) {
+                statsTables.operationFinished(mergeOperationId, Exceptions.messageOf(t), ramAccountingContext.totalBytes());
+            }
+        });
+    }
+
+    private boolean hasDirectResponse(ExecutionNode[] executionNodes) {
+        for (ExecutionNode executionNode : executionNodes) {
+            if (ExecutionNodes.hasDirectResponseDownstream(executionNode.downstreamNodes())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static Map<String, Collection<ExecutionNode>> groupExecutionNodesByServer(ExecutionNode[] executionNodes) {
+        ArrayListMultimap<String, ExecutionNode> executionNodesGroupedByServer = ArrayListMultimap.create();
+        for (ExecutionNode executionNode : executionNodes) {
+            for (String server : executionNode.executionNodes()) {
+                executionNodesGroupedByServer.put(server, executionNode);
+            }
+        }
+        return executionNodesGroupedByServer.asMap();
+    }
+
+    @Override
+    public void start() {
+        statsTables.operationStarted(mergeOperationId, jobId(), "final merge");
+
+        if (nodesByServer.size() == 0) {
+            pageDownstreamContext.finish();
+            return;
+        }
+
+        int idx = 0;
+        for (Map.Entry<String, Collection<ExecutionNode>> entry : nodesByServer.entrySet()) {
+            String serverNodeId = entry.getKey();
+            Collection<ExecutionNode> executionNodes = entry.getValue();
+
+            if (serverNodeId.equals(TableInfo.NULL_NODE_ID)) {
+                serverNodeId = "_local";
+                // TODO: for this to work the TransportJobAction needs to integrate handlerSideDataCollect operation ?
```

Comment:
yep there is another task to change QAF to use the ExecutionNodesTask which should probably include doing this.